### ISSUE TITLE
Fix FakeView in Rails 6

### ISSUE THREAD

**Issue by kueda** [inaturalist/inaturalist#3264]
Our FakeView class doesn't seem to work under Rails 6 due to some changes to ActiveView. FakeView was never a good idea to begin with, and a lot of what we're using it for (providing access to helpers outside of views, a global util class), can be handled differently in more recent versions of Rails, so we should try to ditch it. IMO, this can be divided thusly:

- [x] **Use [`helpers`](https://apidock.com/rails/ActionController/Helpers/helpers) in controllers.** This was introduced in Rails 5 to do exactly what we're using FakeView for in controllers.
- [x] **Remove it from views.** FakeView is redundant inside of a view, not sure why we have it there at all.
- [ ] **Move helper methods that don't require a view context into models or other utility classes.** For example, `ApplicationHelper#uri_join` probably doesn't need to be in a helper.
- [ ] **Replace URL helpers in models with...** This is where I get a bit stuck. For some things, like saving an image url for indexing or getting a path for cache expiry, URL helpers are really useful. Do we still need some class that can construct URLs and is accessible from models, or are there better ways of doing this?

**Comment by kueda** [comment]
Well, I did the easy stuff. Not sure what to do about uses in models and elsewhere, like DarwinCore. We need some way to render templates and generate URLs outside of a controller/view context, so maybe something like FakeView is an ok way to do it. A lot of the logic in the models seems like it should be elsewhere, just not sure where.

**Comment by todtb** [comment]
Hey @kueda,

Re: 
> Replace URL helpers in models with... This is where I get a bit stuck. For some things, like saving an image url for indexing or getting a path for cache expiry, URL helpers are really useful. Do we still need some class that can construct URLs and is accessible from models, or are there better ways of doing this?

I think `Rails.application.routes.url_helpers` could be of some use here. 

Another pattern inside of controllers is to use `helper_method`
```rb
# app/models/post.rb:162
def to_plain_s(options = {})
    s = self.title || ""
    s += ", by #{self.user.try(:login)}" unless options[:no_user]
    s
end
helper_method :to_plain_s
```
which allows the method to be used in controllers where needed and exposed to the associated views.

Is this still work that is on the table?

**Comment by kueda** [comment]
Yeah, I think this is still on the table. We've been running on Rails 6 for a while now so clearly whatever we did is working, but `FakeView` remains an upgrade liability, so I think any good ways for refactoring `FakeView` out of the models would be a good thing.

**Comment by todtb** [comment]
Looking into this a bit more and going over the work you've already done, I'm actually not sure where you'd want to go with this. If `FakeView` is a wrapper around including `Rails.application.routes.url_helpers`, I think that mostly works for access to the helpers from the models. Sure, methods like `place_geometry_kml_url` could probably get pulled out, but I'm not seeing the issues that were present when this was first opened in 10/21. The only alternative I can think of is to call `Rails.application.routes.url_helpers` directly from the model, but I sense you had something else in mind? What specifically is the liability here you're looking to mitigate?

**Comment by kueda** [comment]
I guess the liability is just that `FakeView` intentionally violates some of the Rails architecture for the sake of convenience, like holding on to an instance of `ApplicationController`, and I think the fact that it required so much work in the Rails 6 upgrade is a sign that it probably shouldn't exist. So it's not like a clear and present danger, I just... don't feel good about it.

Some of the ways we use it definitely look like red flags to me, like `Observation#to_plain_s`, and our `to_plain_s` methods in general. They're convenient, but that level of display logic probably doesn't belong in a model. I'm just not sure where it *does* belong.

Anyway, to your original question: is this still on the table? Kind of, but way over on the other side, beyond arm's reach, about to fall onto the floor.
